### ISSUE TITLE
warning in doc build: The fit may be unsuccessful (modeling tied constraints)

### ISSUE THREAD

**Issue by embray** [astropy/astropy#11369]
### Description

I noticed in the doc build for #11367 it failed due to a likely unrelated warning coming from the modeling package:

```python
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/11367/lib/python3.8/site-packages/scipy/optimize/optimize.py:282: RuntimeWarning: Values in x were outside bounds during a minimize step, clipping to bounds
  warnings.warn("Values in x were outside bounds during a "
/home/docs/checkouts/readthedocs.org/user_builds/astropy/envs/11367/lib/python3.8/site-packages/astropy/modeling/fitting.py:1166: AstropyUserWarning: The fit may be unsuccessful; check fit_info['message'] for more information.
  warnings.warn("The fit may be unsuccessful; check "
```

I thought it might be a fluke, but locally I get the same warning, as well as

```
/home/embray/src/astropy/astropy/astropy/timeseries/__init__.py:docstring of astropy.timeseries.BoxLeastSquaresResults.popitem:: WARNING: py:class reference target not found: (k, v), remove and return some (key, value) pair as a
```

There is something fishy going on here.

### System Details

Linux-5.4.0-65-generic-x86_64-with-Ubuntu-18.04-bionic
Python 3.7.5 (default, Nov  7 2019, 10:50:52) 
[GCC 8.3.0]
Numpy 1.19.4
astropy 4.3.dev644+gbb12bf9b2
Scipy 1.5.4
Matplotlib 3.3.3

**Comment by embray** [comment]
The `docstring of astropy.timeseries.BoxLeastSquaresResults.popitem:: WARNING: py:class reference target not found: (k, v), remove and return some (key, value) pair as a` warning, which I got locally but did *not* see on RTD is weird.

The `BoxLeastSquaresResults` is just inheriting `popitem` from its `dict` base class (I think it is weird that this uses `dict` but maybe there is a good reason for that).

Nevertheless, the docstring for popitem is:

```python
>>> from astropy.timeseries import BoxLeastSquaresResults
>>> print(BoxLeastSquaresResults.popitem.__doc__)
D.popitem() -> (k, v), remove and return some (key, value) pair as a
2-tuple; but raise KeyError if D is empty.
```

So why does Spinx think that "(k, v), remove and return some (key, value) pair as a" is a class reference target?

**Comment by embray** [comment]
It looks like in Python 3.8 they changed the formatting of this docstring:

```python
>>> print(dict.popitem.__doc__)
Remove and return a (key, value) pair as a 2-tuple.

Pairs are returned in LIFO (last-in, first-out) order.
Raises KeyError if the dict is empty.
```

Does Sphinx have a syntax I don't know about where any time it sees an `->` in the docstring it should be interpreted as a function signature with annotation?

**Comment by embray** [comment]
I think this warning was not actually the reason the doc build for #11367 failed.  In that case I had a misspelled reference.  Nevertheless I'm curious about the warning.  It traced it to coming from the plot build for this example:

https://docs.astropy.org/en/stable/modeling/example-fitting-constraints.html#tied-constraints

I think in this case the warning can just be ignored.  I think it happens as the fitter struggles to adjust the fit better to the high spikes of those spectral lines.

**Comment by pllim** [comment]
Okay, so `py:class reference target not found` warning is gone, right?

**Comment by pllim** [comment]
We can either ignore the warning (you cannot use doctest directive inside plot directive, so you'll have to use the ugly `catch_warnings` approach) or modify the example to not give a warning.

**Comment by mhvk** [comment]
@pllim pointed me to this - it would be nice to get rid of those unnecessary warnings as they hide real problems (see #12541 - thanks @pllim for avoiding a merge with an actual docs problem in it!).

Obviously, the ideal solution is to modify the example such that it does not give any warnings...  But frankly I'd be in favour of in the meantime just skipping the example altogether...

**Comment by pllim** [comment]
That modeling example is not the only one that emits these warnings. There are a bunch of FITS warnings now too. It is hard to tell where they are coming from on the RTD log. I wonder if a more generic solution (maybe in `sphinx-astropy`?) is needed to ignore these...

**Comment by saimn** [comment]
Putting a `traceback.print_stack()` in the `did not parse as fits unit` unit warning confirms that those uncatched warnings come from the plots (`matplotlib.sphinxext.plot_directive`).

**Comment by pllim** [comment]
Ideas from 2021-12-16 infrastructure tag-up:

* If we cannot find an easy way to fix this, we can document how to read RTD log (e.g., only look for capital “WARNING”).
* Add hidden config sections into narrative docs that suppress the warnings coming from data fitting, etc.

**Comment by pllim** [comment]
@dhomeier fixed the existing warnings in #12788 . Hopefully we are more vigilant going forward so they don't creep back in. Thanks, Derek!
